connect/runtime/src/main/java/org/apache/kafka/connect/runtime/errors/DeadLetterQueueReporter.java

```diff
@@ -0,0 +1,153 @@
+    private KafkaProducer<byte[], byte[]> kafkaProducer;
+    private ErrorHandlingMetrics errorHandlingMetrics;
+
+    private static final ConfigDef CONFIG_DEF = new ConfigDef()
+            .define(DLQ_TOPIC_NAME, ConfigDef.Type.STRING, DLQ_TOPIC_DEFAULT, ConfigDef.Importance.HIGH, DLQ_TOPIC_NAME_DOC);
+
+    public static DeadLetterQueueReporter createAndSetup(WorkerConfig workerConfig,
+                                                         ConnectorConfig connConfig, Map<String, Object> producerProps) {
+        String topic = connConfig.getString(PREFIX + "." + DLQ_TOPIC_NAME);
+
+        try (AdminClient admin = AdminClient.create(workerConfig.originals())) {
+            if (!admin.listTopics().names().get().contains(topic)) {
+                log.error("Topic {} doesn't exist. Will attempt to create topic.", topic);
+                NewTopic schemaTopicRequest = new NewTopic(topic, DLQ_NUM_DESIRED_PARTITIONS, DLQ_MAX_DESIRED_REPLICATION_FACTOR);
+                admin.createTopics(singleton(schemaTopicRequest)).all().get();
+            }
+        } catch (InterruptedException e) {
+            throw new ConnectException("Could not initialize dead letter queue with topic=" + topic, e);
+        } catch (ExecutionException e) {
+            if (!(e.getCause() instanceof TopicExistsException)) {
+                throw new ConnectException("Could not initialize dead letter queue with topic=" + topic, e);
+            }
+        }
+
+        KafkaProducer<byte[], byte[]> dlqProducer = new KafkaProducer<>(producerProps);
+        return new DeadLetterQueueReporter(dlqProducer);
+    }
+
+    /**
+     * Initialize the dead letter queue reporter with a {@link KafkaProducer}.
+     *
+     * @param kafkaProducer a Kafka Producer to produce the original consumed records.
+     */
+    // Visible for testing
+    DeadLetterQueueReporter(KafkaProducer<byte[], byte[]> kafkaProducer) {
+        this.kafkaProducer = kafkaProducer;
+    }
+
+    @Override
+    public void configure(Map<String, ?> configs) {
+        config = new DeadLetterQueueReporterConfig(configs);
+    }
+
+    @Override
+    public void setMetrics(ErrorHandlingMetrics errorHandlingMetrics) {
+        this.errorHandlingMetrics = errorHandlingMetrics;
+    }
+
+    /**
+     * Write the raw records into a Kafka topic.
+     *
+     * @param context processing context containing the raw record at {@link ProcessingContext#consumerRecord()}.
+     */
+    public void report(ProcessingContext context) {
+        if (config.topic().isEmpty()) {
+            return;
+        }
+
+        errorHandlingMetrics.recordDeadLetterQueueProduceRequest();
+
+        ConsumerRecord<byte[], byte[]> originalMessage = context.consumerRecord();
+        if (originalMessage == null) {
+            errorHandlingMetrics.recordDeadLetterQueueProduceFailed();
+            return;
+        }
+
+        ProducerRecord<byte[], byte[]> producerRecord;
+        if (originalMessage.timestamp() > 0) {
```

Comment:
This seems error prone to be checking for this, rather than using the ConsumerRecord's timestamp type.

Reply:
Agreed here, can we use https://github.com/apache/kafka/blob/trunk/clients/src/main/java/org/apache/kafka/common/record/RecordBatch.java#L47 here to be specific to the one sentinel value we have? That would also make it robust to changes for https://cwiki.apache.org/confluence/display/KAFKA/KIP-228+Negative+record+timestamp+support (maybe not robust to every possible fix, but at least for ones that just reserve that one timestamp value as a sentinel value).

Reply:
Yeah, that works, too, and is more align with the current code.